lcls-twincat-common-components/Library/Devices/PPM/FB_PPM_PowerMeter.TcPOU

```diff
@@ -28,61 +40,110 @@ VAR
     fVoltageBuffer: ARRAY[1..1000] OF LREAL;
 
     {attribute 'pytmc' := '
-        pv: CALIB
+        pv: STC
         io: input
     '}
-    fCalibBase: LREAL;
+    fbTempSensor: FB_CC_TempSensor;
 
     {attribute 'pytmc' := '
-        pv: CALIB_BUFFER
-        io: input
+        pv: AccTypeFromEpics
+        link: TPR:FEES:GD02:0:MODE.RVAL
     '}
-    fCalibBaseBuffer: ARRAY[1..1000] OF LREAL;
+    uAccelerator: UINT;
 
     {attribute 'pytmc' := '
-        pv: MJ
-        io: input
+        pv: NCBeamRateFromEpics
+        link: TPG:SYS0:1:DST04:RATE
+    '}
+    uNCFrequency: UINT;
+
+    {attribute 'pytmc' := '
+        pv: SCBeamRateFromEpics
+        link: TPG:SYS0:1:DST04:RATE
+    '}
+    uSCFrequency: UINT;
+
+    uFrequency: UINT;
+
+    // Pulse-by-pulse energy
+    {attribute 'pytmc' := '
+        pv: PulseEnergy
+        io: i
         field: EGU mJ
     '}
     fCalibMJ: LREAL;
 
+    // Pulse-by-pulse energy buffer
     {attribute 'pytmc' := '
-        pv: MJ_BUFFER
-        io: input
+        pv: PulseEnergy_BUFFER
+        io: i
         field: EGU mJ
     '}
     fCalibMJBuffer: ARRAY[1..1000] OF LREAL;
 
+    // Wattage
     {attribute 'pytmc' := '
-        pv: STC
-        io: input
+        pv: PulseWattage
+        io: i
+        field: EGU W
     '}
-    fbTempSensor: FB_CC_TempSensor;
+    fPulseWattage: LREAL;
 
+    // Background voltage
     {attribute 'pytmc' := '
-        pv: CALIB:OFFSET
+        pv: BackgroundVoltage
         io: io
+        field: EGU V
     '}
-    fCalibRelOffset: LREAL;
+    fBackgroundVoltage: LREAL;
+
+    fBackgroundVoltageAcc: LREAL;
+    fBackgroundVoltageSum: LREAL;
+    fBackgroundVoltageStale: LREAL;
+    tonBackgroundAutoCollecting: TON;
+    fBackgroundVoltageBuffer: ARRAY[1..100] OF LREAL;
+    udBackgroundVoltageBufferIndex: UDINT := 1;
+    i: UDINT;
 
     {attribute 'pytmc' := '
-        pv: CALIB:RATIO
+        pv: BackgroundMode
         io: io
     '}
-    fCalibRelRatio: LREAL;
+    BACKGROUND_MODE: (Manual, Passive) := Passive;
+
+    // Boolean to trigger collection of background voltages
+    {attribute 'pytmc' := '
+        pv: BackgroundCollect
+        io: i
+    '}
+    bBackgroundCollect: BOOL;
 
     {attribute 'pytmc' := '
-        pv: CALIB:MJ_RATIO
+        pv: BackgroundCollectionTime
         io: io
+        field: EGU s
     '}
-    fCalibMJRatio: LREAL;
+    // Time in seconds to collect background voltages for
+    uBackgroundCollectionTime: UINT := 60;
+
+    tofBackgroundCollecting: TOF;
+    udBackgroundManualCount: UDINT;
 
     fbGetPMVoltage: FB_AnalogInput;
+
     fbVoltageBuffer: FB_LREALBuffer;
-    fbCalibBaseBuffer: FB_LREALBuffer;
     fbCalibMJBuffer: FB_LREALBuffer;
+
+    FFO: FB_FastFault :=(
+        i_Desc := 'Fault occurs when the temprature trip point is reached',
+        i_TypeCode := 16#f400);
+    bOverAllowableEnergy: BOOL;
 END_VAR
-]]></Declaration>
+VAR CONSTANT
+    fWavelengthEdge1: REAL := 300 + 1.0E-44;
```

Comment:
Is this 1.0E-44 related to the smallest increment of a REAL?
Is there a reason these can't be integers?

Reply:
It is and there is not.